contracts/mirror/src/contract.rs

```diff
@@ -0,0 +1,473 @@
+            expected: BondState::Bonded.to_string(),
+            got: bond.state.to_string(),
+        }
+    );
+    let backup = bond.backup.clone().ok_or(ContractError::BackupIsNotSet)?;
+
+    ensure_eq!(info.sender, backup, ContractError::Unauthorized);
+    bond.return_type = return_type.clone();
+    BONDS.save(deps.storage, id, &bond)?;
+    let attrs = vec![
+        attr("action", "change_return_type"),
+        attr("id", id.to_string()),
+        attr("return_type", return_type.to_string()),
+    ];
+    Ok(response("change_return_type", CONTRACT_NAME, attrs))
+}
+
+pub fn execute_update_bond(
+    deps: DepsMut<NeutronQuery>,
+    _env: Env,
+    info: MessageInfo,
+    id: u64,
+    receiver: String,
+    backup: Option<String>,
+    return_type: ReturnType,
+) -> ContractResult<Response<NeutronMsg>> {
+    cw_ownable::assert_owner(deps.storage, &info.sender)?;
+    let mut bond = BONDS.load(deps.storage, id)?;
+    bond.receiver = receiver.clone();
+    bond.backup = backup
+        .clone()
+        .map(|a| deps.api.addr_validate(&a))
+        .transpose()?;
+    bond.return_type = return_type.clone();
+    BONDS.save(deps.storage, id, &bond)?;
+    let attrs = vec![
+        attr("action", "update_bond"),
+        attr("id", id.to_string()),
+        attr("receiver", receiver),
+        attr("backup", backup.unwrap_or_default()),
+        attr("return_type", return_type.to_string()),
+    ];
+    Ok(response("update_bond_state", CONTRACT_NAME, attrs))
+}
+
+pub fn execute_bond(
+    deps: DepsMut<NeutronQuery>,
+    _env: Env,
+    info: MessageInfo,
+    receiver: String,
+    r#ref: Option<String>,
+    backup: Option<String>,
+) -> ContractResult<Response<NeutronMsg>> {
+    let Config {
+        core_contract,
+        prefix,
+        ..
+    } = CONFIG.load(deps.storage)?;
+    let backup = backup
+        .as_ref()
+        .map(|addr| deps.api.addr_validate(addr))
+        .transpose()?;
+    bech32::decode(&receiver).map_err(|_| ContractError::WrongReceiverAddress)?;
+    ensure!(receiver.starts_with(&prefix), ContractError::InvalidPrefix);
+    let coin = cw_utils::one_coin(&info)?;
+    let counter = COUNTER.load(deps.storage)?;
+    let id = counter + 1;
+    COUNTER.save(deps.storage, &id)?;
+    let attrs = vec![
+        attr("action", "bond"),
+        attr("id", id.to_string()),
+        attr("receiver", receiver.to_string()),
+        attr("ref", r#ref.clone().unwrap_or_default()),
+        attr(
+            "backup",
+            backup.as_ref().map(|x| x.to_string()).unwrap_or_default(),
+        ),
+    ];
+    BONDS.save(
+        deps.storage,
+        id,
+        &BondItem {
+            receiver,
+            backup,
+            received: None,
+            amount: coin.amount,
+            return_type: drop_staking_base::state::mirror::ReturnType::default(),
+            state: drop_staking_base::state::mirror::BondState::default(),
+        },
+    )?;
+    let msg = SubMsg::reply_on_success(
```

Comment:
what if Bond operation on the core contract fails what happens with attached funds? Looks like mirror contract will silently ignores failure and mark this call as succeeded.

Reply:
the tx is reverted if it fails

Reply:
even if submessage fails?

Reply:
Sorry, it was my misunderstanding how it works, found this explanation https://github.com/CosmWasm/cosmwasm/blob/main/SEMANTICS.md#submessages